ksp_plugin/interface.hpp

```diff
@@ -25,13 +26,45 @@ using ksp_plugin::World;
 
 namespace interface {
 
-struct LineAndIterator {
-  explicit LineAndIterator(RenderedTrajectory<World> const& rendered_trajectory)
-      : rendered_trajectory(rendered_trajectory) {}
-  RenderedTrajectory<World> const rendered_trajectory;
-  RenderedTrajectory<World>::const_iterator it;
+// A wrapper for a container and an iterator into that container.
+class Iterator {
+ public:
+  virtual ~Iterator() = default;
+
+  virtual bool AtEnd() const = 0;
+  virtual void Increment() = 0;
+  virtual int Size() const = 0;
 };
 
+// A concrete, typed subclass of |Iterator|.  It holds a |Container| of elements
+// of type |Internal|.
+template<typename Internal, template<typename...> class Container>
```

Comment:
Remove the template parameter `Internal`, and use `Container::value_type` (from the concept [`Container`](http://en.cppreference.com/w/cpp/concept/Container)) instead.
This should allow for type inference on the constructor, turning
```cpp
*periapsides = new TypedIterator<LineSegment<World>, std::vector>(std::move(rendered_periapsides));
```
into

```cpp
*periapsides = new TypedIterator(std::move(rendered_periapsides));
```